client_plugin/drivers/vfile/vfile_driver.go

```diff
@@ -698,6 +745,43 @@ func (d *VolumeDriver) UnmountVolume(name string) error {
 		return err
 	}
 
+	// remove VM from ClientList
+	nodeID, addr, _, err := d.dockerOps.GetSwarmInfo()
+	if err != nil {
+		log.WithFields(
+			log.Fields{"volume name": name,
+				"error": err,
+			}).Error("Failed to get IP address from docker swarm ")
+		return err
+	}
+
+	err = d.removeVMFromClientList(name, nodeID, addr)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"volume name": name,
+				"error": err,
+			}).Error("Failed to remove VM IP from ClientList")
+		return err
+	}
+	mountpoint := d.GetMountPoint(name)
+	err = fs.Unmount(mountpoint)
```

Comment:
Can you explain why we cannot unmount first and then decrement the gref? 

Otherwise the code looks good. The error handling is not done correctly and it seems like we need to think a bit more on it. This can be done as a different code change.

Reply:
We can do that.  Current code is just to make sure it consistent with the mount workflow (increase gref and then do mount). Is there any reason you think we should do umount first and then decrement the gref?

Reply:
As we discussed, the auto decrease of gref(1->0) will trigger deletion of the SMB container and that could fail since it is mounted/used by another container. If manager node gets to delete the SMB service successfully and if the worker node fails unmounting, we are increasing the gref and will lead to inconsistent state.

As I mentioned, the error handling needs to be thought through properly since there are different parties involved (manager nodes, worker nodes) and different failures (etcd failure, SMB container deploy failure, mount/unmount failure etc).

Reply:
@SandeepPissay I addressed your comments to unmount first and then decrease gref. Please reivew it.